clang/lib/CConv/ConstraintVariables.cpp

```diff
@@ -1581,7 +1581,12 @@ void PointerVariableConstraint::brainTransplant(ConstraintVariable *FromCV,
 }
 
 void PointerVariableConstraint::mergeDeclaration(ConstraintVariable *FromCV,
-                                                 ProgramInfo &Info) {
+                                                 ProgramInfo &Info,
+                                                 std::string &ReasonFailed) {
+  if (FromCV->getOriginalTy() != this->getOriginalTy()) {
```

Comment:
I think this check is probably too strong, which is why you are getting regressions. None of the existing tests should fail. The existing types *can* be different, but only if they are *compatible.* What `mergeDeclaration` is doing is taking two prototypes that are compatible and merging them together into one. 

Compatibility is a loose concept in my mind, but it basically boils down to the two functions having the same number of arguments and the same kinds of pointer for each, with the only difference whether one might have an `itype` where the other doesn't, or one might have a checked type and the other an unchecked one (e.g., `_Ptr<int>` vs. `int *`). 

Maybe @aaronjeline can comment based on what he knows. 

It might be that this is not the right place to put this check, i.e., that you should interleave the check in with the merging process, and check for \# parameters, types, etc. as you go and then fail there.

Reply:
The possibility that the tool might be run on a file that already has Checked C types in it never crossed my mind, so that's definitely a vulnerability I wasn't accounting for. At any rate, I also think this is too strict, so I'll do a bit more digging later to find a way to get it to be robust. 
I know you said compatibility is a loose concept for you, but if you (or anyone reading this right now) could answer a couple of questions, it would give me some direction on how to handle this: 
1) Is it the case that if the base types are different, we should always fail? (i.e. char vs int. These can be pointers, be _Ptrs, or whatever they want to be. But if we observe that one of them is an int and another is a char, is there ever a case in which we want to allow it to happen? If there is no such case, we can detect their base types, so that simplifies things a bit. (Although not by much, because it's possible to have typedefs or macros that are compatible i.e. size_t vs uint and it's not apparent to me that the BaseType calculation in our code unfolds them) 
2) Is it always the case that different pointers should be rejected (by this I mean, if the i-th argument in two functions are int** and int* respectively)?

Reply:
The way you can check some of these things is to make two prototypes and put them in the same file and run the Checked C `clang` (not our tool, the compiler) and see what happens. If it gives you an error, then this is bad. If it says nothing, good. If a warning, also "good." For example, for your last question, if I use this file:
```
void foo(int *x);
void foo(int **x);
```
I get this outcome:
```
a.c:2:6: error: conflicting types for 'foo'
void foo(int **x);
     ^
a.c:1:6: note: previous declaration is here
void foo(int *x);
     ^
1 error generated.
```
whereas if I do this:
```
void foo(int *x);
void foo(int *x: itype(_Ptr<int>));
```
The compiler is happy with it.

Reply:
BTW, to see if a Checked C type is compatible with a legacy type, you can do put them both in an `itype`. For example, to see if `int **` is compatible with `_Ptr<int>`, you can write:
```
void foo(int **x: itype(_Ptr<int>));
```
Then pass this to Checked C `clang`. It doesn't like it, as we would expect:
```
a.c:2:25: error: mismatch between interface type '_Ptr<int>' and declared type
      'int **'
void foo(int **x: itype(_Ptr<int>));
                        ^
1 error generated.
```

Reply:
Oh, this is a game changer in terms of debugging by cases. Thanks! 

For right now, at least on the int** vs int* debacle, the difference appears to be pretty clear. (In that if you do getCVars() for both of them, the number of CVars for the first will be two (one to represent the outer pointer, one to represent dereferencing the outer pointer to get the inner pointer), and the number of CVars for the second will be 1 (there's only one pointer)).
I'll probably have to do a bit of a code review in addition to looking at the Checked C typechecker to ensure that the above is an invariant (i.e. different lengths iff incompatible).

Reply:
I think that's probably an invariant. @jackastner might be able to comment on that. For a while it wasn't, but then I think we decided to try to make it one, but not sure if we finished that work.